File: javascript/smartlist/smartlist-api-mysql-project/src/modules/candidate/candidate-question-strategy/base-question-strategy.ts
import { AnswerOption } from 'src/entities/answer-option.entity';
import { Question } from 'src/entities/question.entity';
import { CandidateResponseCodes } from 'src/modules/candidate/candidate.response.codes';
import { SubmitAnswerReqDto } from 'src/modules/candidate/dto/request/submit-answer.req.dto';
import { AnswerOptionRepository } from 'src/modules/question/answer-option.repository';
import { arrayHasDuplicates } from 'src/utils/app.utils';
import { getCustomRepository } from 'typeorm';
import { IQuestionStrategy } from '../interfaces/interface-question-strategy';

export abstract class BaseQuestionStrategy implements IQuestionStrategy {
  protected answerOptionRepository: AnswerOptionRepository;
  constructor(
    protected submitAnswerReqDto: SubmitAnswerReqDto,
    protected question: Question,
  ) {
    this.answerOptionRepository = getCustomRepository(AnswerOptionRepository);
  }

  async validate() {
    if (arrayHasDuplicates(this.submitAnswerReqDto.answer_ids)) {
      throw CandidateResponseCodes.INVALID_ANSWER_ID;
    }

    if (this.submitAnswerReqDto.answer_ids.length == 0) {
      throw CandidateResponseCodes.INVALID_ANSWER_ID;
    }

    for (let i = 0; i < this.submitAnswerReqDto.answer_ids.length; i++) {
      const answerOption: AnswerOption =
        await this.answerOptionRepository.findByIdWithQuestion(
          this.submitAnswerReqDto.answer_ids[i],
        );

      if (
        answerOption == null ||
        answerOption.question.id != this.question.id
      ) {
        throw CandidateResponseCodes.INVALID_ANSWER_ID;
      }
    }
  }

  abstract score(): number;
}

File: java/smartlist/src/main/java/com/smartlist/module/candidate/questionstrategy/BaseQuestionStrategy.java
package com.smartlist.module.candidate.questionstrategy;

import com.smartlist.module.candidate.interfaces.IQuestionStrategy;
import com.smartlist.model.AnswerOption;
import com.smartlist.model.Question;
import com.smartlist.module.candidate.CandidateResponseCodes;
import com.smartlist.module.candidate.dto.request.SubmitAnswerReqDTO;
import com.smartlist.module.question.AnswerOptionRepository;
import com.smartlist.utils.AppUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.server.ResponseStatusException;

import java.util.Objects;
import java.util.Optional;

public abstract class BaseQuestionStrategy implements IQuestionStrategy {

    protected final AnswerOptionRepository answerOptionRepository;
    protected final SubmitAnswerReqDTO submitAnswerReqDto;
    protected final Question question;

    protected BaseQuestionStrategy(final SubmitAnswerReqDTO submitAnswerReqDto, final Question question, final AnswerOptionRepository answerOptionRepository) {
        this.submitAnswerReqDto= submitAnswerReqDto;
        this.question = question;
        this.answerOptionRepository = answerOptionRepository;
    }

    public void validate() {
        if (AppUtils.arrayHasDuplicates(this.submitAnswerReqDto.getAnswerIds())) {
            throw CandidateResponseCodes.INVALID_ANSWER_ID;
        }

        for (int i = 0; i < this.submitAnswerReqDto.getAnswerIds().size(); i++) {
            Optional<AnswerOption> optionalAnswerOption = this.answerOptionRepository.findById(this.submitAnswerReqDto.getAnswerIds().get(i));
            AnswerOption answerOption =  optionalAnswerOption.orElseThrow(this::throwIfAnswerOptionInvalid);

            if (!Objects.equals(answerOption.getQuestion().getId(), this.question.getId())) {
                throw CandidateResponseCodes.INVALID_ANSWER_ID;
            }
        }
    }

    private ResponseStatusException throwIfAnswerOptionInvalid() {
        return CandidateResponseCodes.INVALID_ANSWER_ID;
    }
}
